### PR TITLE
harmonize distance unit for 4 nodes  buffer, multiple ring buffer, nearest join, simplify

### PR DESCRIPTION
**Pull request by UrbanGISer**
all 4 nodes were revised, buffer, multiple ring buffer, nearest join, simplify. The length node use a class object which might leave to you :)

### CODE REVIEW

File: knime_extension/src/nodes/spatialtool.py [hunk 1]
```diff
@@ -806,8 +886,10 @@ def execute(self, exec_context: knext.ExecutionContext, input_1):
                 gdf0 = gp.overlay(gdf0, ci, how="union")
         # Add ring radius values as a new column
         gdf0["dist"] = bufferlist
+        # convert it back to the orginal CRS of input data
+        gdf0.to_crs(origincrs, inplace=True)
         gdf0 = gdf0.reset_index(drop=True)
-        exec_context.set_progress(0.1, "Buffering done")
+        gdf0 = gdf0.to_crs(origincrs)
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:892]: revised

File: knime_extension/src/nodes/spatialtool.py [hunk 2]
```diff
@@ -869,12 +962,33 @@ def configure(self, configure_context, input_schema_1):
...
+            newcrs = gdf.crs
+        gdf = gdf.to_crs(newcrs)
+
+        if self.unitset == "Kilometer":
+            newsimplifydist = self.simplifydist * 1000
+        else:
+            newsimplifydist = self.simplifydist
+
+        gdf[self.geo_col] = gdf.geometry.simplify(newsimplifydist)
+        gdf = gdf.to_crs(crsinput)
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:989]: revised

File: knime_extension/src/nodes/spatialtool.py [hunk 3]
```diff
@@ -770,24 +840,34 @@ def configure(self, configure_context, input_schema_1):
...
+            )
+            newcrs = 3857
+        elif crsinput.is_projected and self.bufferunit == "Degree":
+            exec_context.set_warning(
+                "CRS unit does not match, CRS is converted to  default CRS epsg:4326 "
+            )
+            newcrs = 4326
+        else:
+            newcrs = gdf.crs
+        gdf = gdf.to_crs(newcrs)
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:861]: revised

File: knime_extension/src/nodes/spatialtool.py [hunk 4]
```diff
@@ -89,16 +96,26 @@ def execute(self, exec_context: knext.ExecutionContext, input):
...
+
+        if self.unitset == "Kilometer":
+            newbufferdist = self.bufferdist * 1000
+        else:
+            newbufferdist = self.bufferdist
+        gdf["geometry"] = gdf["geometry"] = gdf.buffer(newbufferdist)
+        gdf = gdf.to_crs(crsinput)
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:118]: revised

File: knime_extension/src/nodes/spatialtool.py [hunk 5]
```diff
@@ -403,21 +429,65 @@ def execute(self, exec_context: knext.ExecutionContext, left_input, right_input)
...
+            maxdistvalue = self.maxdist
+
         gdf = gp.sjoin_nearest(
             left_gdf,
             right_gdf,
             how=self.join_mode.lower(),
-            max_distance=self.maxdist,
-            distance_col="NearDist",
+            max_distance=maxdistvalue,
+            distance_col="neardist",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:465]: only neardist will be output, revised it back to NearDist

File: knime_extension/src/nodes/spatialtool.py [hunk 6]
```diff
@@ -403,21 +429,65 @@ def execute(self, exec_context: knext.ExecutionContext, left_input, right_input)
...
             lsuffix="1",
             rsuffix="2",
         )
+        if gdf.shape[0] < 1:
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:469]: Normally, returning an empty table is meaningless, I keep this

File: knime_extension/src/nodes/spatialtool.py [hunk 7]
```diff
@@ -728,7 +798,7 @@ def execute(self, exec_context: knext.ExecutionContext, left_input, right_input)
     description="""This node generate multiple polygons with a series distances of each geometric object.
 
 **Note:** If the input table contains multiple rows the node first computes the union of all geometries before 
-computing the buffers from the union.
+computing the buffers from the union. This node only accepts buffer distances based on a projected CRS.
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:801]: deleted, as it is a previous function

File: knime_extension/src/nodes/spatialtool.py [hunk 8]
```diff
@@ -89,16 +96,26 @@ def execute(self, exec_context: knext.ExecutionContext, input):
...
-        gdf[knut.get_unique_column_name("geometry", input.schema)] = gdf.buffer(
-            self.bufferdist
-        )
-        exec_context.set_progress(0.1, "Buffering done")
-        LOGGER.debug(
-            "Feature geometry " + self.geo_col + " buffered with" + str(self.bufferdist)
-        )
+        if crsinput.is_geographic and self.unitset != "Degree":
+            exec_context.set_warning(
+                "CRS unit does not match, CRS is converted to default CRS epsg:3857 "
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:101]: adpated

File: knime_extension/src/nodes/spatialtool.py [hunk 9]
```diff
@@ -89,16 +96,26 @@ def execute(self, exec_context: knext.ExecutionContext, input):
...
+            )
+            newcrs = 3857
+        elif crsinput.is_projected and self.unitset == "Degree":
+            exec_context.set_warning(
+                "CRS unit does not match, CRS is converted to  default CRS epsg:4326 "
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialtool.py:106]: the  statement in all other nodes were changed as we;;

**Comment by koettert** [comment]
Please check how we can distinguish between projections that use degree, meter and **feet**.

**Comment by koettert** [comment]
Obsolete with https://github.com/spatial-data-lab/knime-geospatial-extension/pull/191

**Pull request closed without merging**